ZMQ: Small cleanups in the ZMQ code

domob1812 opened the pull request:
This contains various small code cleanups that make the ZMQ code easier to read and maintain (at least in my opinion).  The only functional change is that a potential memory leak is fixed that would have occured when a notifier is removed from the `notifiers` list after its callback function returned `false` (which is likely not relevant in practice but still a bug).

domob1812 committed:
Use std::unique_ptr for CZMQNotifierFactory.

Instead of returning a raw pointer from CZMQNotifierFactory and
implicitly requiring the caller to know that it has to take ownership,
return a std::unique_ptr to make this explicit.

This also changes the typedef for CZMQNotifierFactory to use the new
C++11 using syntax, which makes it (a little) less cryptic.

domob1812 committed:
Replace zmqconfig.h by a simple zmqutil.

zmqconfig.h is currently not really needed anywhere, except that
it declares zmqError (which is then defined in
zmqnotificationinterface.cpp).  Note in particular that there is
no need to conditionally include zmq.h only if ZMQ is enabled, because
the place in the core code where the ZMQ library itself is included
(init.cpp) is conditional already on that.

This commit removes zmqconfig.h and replaces it by a much simpler
zmqutil.h library for zmqError.  The definition of the function is
moved to the matching (newly created) zmqutil.cpp.

domob1812 commented:
I split the patch into three commits which I think make sense logically and for the review.  But of course I'm happy to squash any of the changes, or to revert parts of them (as code readability is a matter of taste).

promag commented:
Concept ACK.

From a quick view LGTM.

domob1812 commented:
Rebased

kostyantyn commented:
What do you think about wrapping the code with `zmq` namespace and removing `ZMQ/zmq` prefix from class/function names? 
Basically when the location of the file is `a/b/c.cpp` then we have:
```c++
namespace a {
namespace b {
class C {}
}
}
```

domob1812 commented:
@kostyantyn:  At least for now, there seems to be not much namespace usage at all in the Bitcoin code - so I'm not sure if we should change that lightly.  Also, note that `zmq` is the namespace that is also used by ZMQ's own C++ language bindings; they are currently not used, but I proposed to change that (#13957).

Also, note that C++ namespaces are [different from Java's](https://stackoverflow.com/questions/2108172/c-namespaces-comparison-to-java-packages), so that some people advise against nesting (e.g. exact correspondence to the directory structure as you suggest).

laanwj commented:
Concept ACK

@jmcorgan do you mind taking a look here?

ken2812221 commented on src/zmq/zmqabstractnotifier.h:
```diff
@@ -5,12 +5,13 @@
 
-#include <zmq/zmqconfig.h>
+#include <memory>
 
```
Missing a bunch of includes here. See appveyor CI result.

domob1812 replied:
I missed that comment - I've now revised the commit to include `<string>`.  There are still CI errors, though, which I don't understand so far.  Has anyone an insight of where they come from?  I'll try to resolve them now.

practicalswift commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -4,79 +4,69 @@
     std::list<const CZMQAbstractNotifier*> result;
-    for (const auto* n : notifiers) {
-        result.push_back(n);
+    for (const auto& n : notifiers) {
```
Could use `std::transform`? :-)

domob1812 committed:
scripted-diff: Rename SendMessage to SendZmqMessage.

Windows headers define SendMessage as a macro, which leads to problems
with the method name "SendMessage".  To circumvent this, we rename the
method to "SendZmqMessage".

-BEGIN VERIFY SCRIPT-
sed -i 's/SendMessage/SendZmqMessage/g' src/zmq/zmqpublishnotifier.*
-END VERIFY SCRIPT-

domob1812 commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -2,14 +2,25 @@
+
+#include <cstdarg>
+#include <cstddef>
+
```
Thanks, very interesting!  From a quick look, I didn't see what my PR changes here and why it worked before - but I guess a defensive rename of the method is not a bad idea.  I did that now in a separate commit and pushed it, just to see if it fixes the CI issues.  If it does, I can include this directly in one of the commits or perhaps just keep it as it is.

domob1812 replied:
Indeed, looks like the tests pass now.

hebasto commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -4,79 +4,69 @@
 #include <zmq/zmqpublishnotifier.h>
+#include <zmq/zmqutil.h>
+
+#include <zmq.h>
```
Why is a guard dropped?
```
#ifdef ENABLE_ZMQ
#include <zmq.h>
#endif
```

hebasto reviewed (changes requested):
Concept ACK.

domob1812 commented:
Rebased.  Is there anything missing for me to do here?  I think this is a real improvement to the quality of the ZMQ code.

fanquake commented:
@jmcorgan Are you able to take a look at this?

domob1812 commented:
Rebased.  Can I do anything to speed up review / merging of this one?  It should not need review by the original author or a ZMQ expert, as this is mostly just "standard" refactoring from manual memory management to the use of proper RAII (e.g. `std::unique_ptr`).

promag commented on src/zmq/zmqabstractnotifier.h:
```diff
@@ -21,9 +23,9 @@ class CZMQAbstractNotifier
+    static std::unique_ptr<CZMQAbstractNotifier> Create()
     {
-        return new T();
+        return std::unique_ptr<CZMQAbstractNotifier>(new T());
```
d10f274a6e0291e752e25b2bd99c0472d22a420f

`MakeUnique`?

domob1812 replied:
Done.

promag reviewed (commented):
Code review ACK ba8b67aee28623d7fdd9b2053bc1784518a4a8ca, with some comments.

domob1812 commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -136,45 +115,44 @@ void CZMQNotificationInterface::Shutdown()
 }
 
-void CZMQNotificationInterface::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload)
+namespace
```
Done.

domob1812 commented:
Thanks for the review!  I've addressed your comments in follow-up commits (I do not want to force-push right now and perhaps mess up your ongoing review).  I will add them as fixups to the proper commits later when you are done with the review.

I can then also split out the creation of `TryForEachAndRemoveFailed` into a separate commit if you prefer that.

domob1812 commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -136,45 +115,44 @@ void CZMQNotificationInterface::Shutdown()
-        else
-        {
+template <typename Function>
+void TryForEachAndRemoveFailed(std::list<std::unique_ptr<CZMQAbstractNotifier>>& notifiers, const Function& func)
```
Yes sure, if you prefer.  Let's finish the review of the code first, and then I will rework the commit history as desired afterwards (else we might mess up the ongoing review with force-pushes).

instagibbs replied:
Note for reviewers: I think this is resolved?

domob1812 commented:
Rebased

hebasto commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -58,23 +51,22 @@ CZMQNotificationInterface* CZMQNotificationInterface::Create()
-        notificationInterface->notifiers = notifiers;
+        notificationInterface.reset(new CZMQNotificationInterface());
+        notificationInterface->notifiers = std::move(notifiers);
 
-        if (!notificationInterface->Initialize())
-        {
-            delete notificationInterface;
-            notificationInterface = nullptr;
+        if (!notificationInterface->Initialize()) {
+            notificationInterface.reset();
         }
     }
 
-    return notificationInterface;
+    return notificationInterface.release();
```
69f92ffce9f0d11d9389610e702b99d9038f3fed

If this code is touched to make it easier to read and maintain, may I suggest a simpler approach:
```
    if (!notifiers.empty()) {
        auto notificationInterface = std::unique_ptr<CZMQNotificationInterface>(new CZMQNotificationInterface());
        notificationInterface->notifiers = std::move(notifiers);

        if (notificationInterface->Initialize()) {
            return notificationInterface.release();
        }
    }

    return nullptr;
```
?

domob1812 replied:
Indeed, looks better.  I've updated the code accordingly (except I changed the `auto notificationInterface = ...` line slightly into what I consider even clearer; but let me know if you prefer your suggestion).

hebasto commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -2,14 +2,25 @@
 
+#include <zmq.h>
+
+#include <cstdarg>
+#include <cstddef>
```
```suggestion
#include <cstdarg>
#include <cstddef>
#include <map>
#include <string>
#include <utility>
```

domob1812 replied:
Indeed, I've pushed this change together with the two nanonits below (in a single commit).

hebasto commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -2,14 +2,25 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
```
Is this required here as no defines from `bitcoin-config.h` is used here?

domob1812 replied:
I guess not, removed.

hebasto commented on src/zmq/zmqutil.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef BITCOIN_ZMQ_ZMQUTIL_H
+#define BITCOIN_ZMQ_ZMQUTIL_H
+
+void zmqError(const char *str);
```
49e2cb78563df27ff0a5d2ad985e074ee98afcce
nanonit:
```suggestion
void zmqError(const char* str);
```

hebasto commented on src/zmq/zmqutil.cpp:
```diff
@@ -0,0 +1,14 @@
+
+#include <zmq.h>
+
+void zmqError(const char *str)
```
49e2cb78563df27ff0a5d2ad985e074ee98afcce
nanonit:
```suggestion
void zmqError(const char* str)
```

hebasto reviewed (changes requested):
Approach ACK 14b40f1b7ce7d0b457e163dbda4e81f5105db753

domob1812 commented:
Thanks for the review, @hebasto, very appreciated!  I've addressed your suggestions now with fixup commits.  I will now rebase this onto current master, and then go clean up the commit history (including splitting out the refactor as suggested by @promag and you).

domob1812 committed:
Various cleanups in zmqnotificationinterface.

This is a pure refactoring of zmqnotificationinterface to make the
code easier to read and maintain.  It replaces explicit iterators
with C++11 for-each loops where appropriate and uses std::unique_ptr
to make memory ownership more explicit.

domob1812 committed:
Simplify and fix notifier removal on error.

This factors out the common logic to run over all ZMQ notifiers, call a
function on them, and remove them from the list if the function fails is
extracted to a helper method.

Note that this also fixes a potential memory leak:  When a notifier was
removed previously after its callback returned false, it would just be
removed from the list without destructing the object.  This is now done
correctly by std::unique_ptr behind the scenes.

domob1812 commented:
Rebase done, and I've now cleaned up the commit history as well (squashed in the "fixup" commits as appropriate).  Also split out the `TryForEachAndRemoveFailed` refactor into its own commit.

hebasto commented:
The first commit 707c4803d5edd2d5ebd6febb33e08bb031630652 is not compiled for me.

domob1812 commented:
> The first commit [707c480](https://github.com/bitcoin/bitcoin/commit/707c4803d5edd2d5ebd6febb33e08bb031630652) is not compiled for me. Make iterator type `auto`?

Sorry for that, it is fixed now.  The overall PR is not changed (except I've rebased it to current `master`), but now each individual commit builds.

hebasto reviewed (changes requested):
ACK 166ddcbef82ba2b6734e8f09584fb73e4f9f98a4.

Mind making the latest commit 166ddcbef82ba2b6734e8f09584fb73e4f9f98a4 a sripted-diff with the following script:
```
sed -i 's/SendMessage/SendZmqMessage/g' src/zmq/zmqpublishnotifier.*
```
?

domob1812 commented:
> Mind making the latest commit [166ddcb](https://github.com/bitcoin/bitcoin/commit/166ddcbef82ba2b6734e8f09584fb73e4f9f98a4) a sripted-diff with the following script:

Good point, done.

instagibbs commented:
utACK 6fe2ef2acb00b1df7f6a0c0dea1a81a1924be0e1

I restarted one travis job that appeared to be a spurious timeout.

hebasto reviewed (approved):
re-ACK 6fe2ef2acb00b1df7f6a0c0dea1a81a1924be0e1, only the latest commit got a scripted-diff since my [previous](https://github.com/bitcoin/bitcoin/pull/13686#pullrequestreview-487649808) review.
